File: src/client/graphics/layers/EmojiTable.ts
import { LitElement, html } from "lit";
import { customElement, state } from "lit/decorators.js";
import { EventBus } from "../../../core/EventBus";
import { AllPlayers } from "../../../core/game/Game";
import { GameView, PlayerView } from "../../../core/game/GameView";
import { TerraNulliusImpl } from "../../../core/game/TerraNulliusImpl";
import { Emoji, emojiTable, flattenedEmojiTable } from "../../../core/Util";
import { CloseViewEvent, ShowEmojiMenuEvent } from "../../InputHandler";
import { SendEmojiIntentEvent } from "../../Transport";
import { TransformHandler } from "../TransformHandler";

@customElement("emoji-table")
export class EmojiTable extends LitElement {
  @state() public isVisible = false;
  public transformHandler: TransformHandler;
  public game: GameView;

  initEventBus(eventBus: EventBus) {
    eventBus.on(ShowEmojiMenuEvent, (e) => {
      this.isVisible = true;
      const cell = this.transformHandler.screenToWorldCoordinates(e.x, e.y);
      if (!this.game.isValidCoord(cell.x, cell.y)) {
        return;
      }

      const tile = this.game.ref(cell.x, cell.y);
      if (!this.game.hasOwner(tile)) {
        return;
      }

      const targetPlayer = this.game.owner(tile);
      // maybe redundant due to owner check but better safe than sorry
      if (targetPlayer instanceof TerraNulliusImpl) {
        return;
      }

      this.showTable((emoji) => {
        const recipient =
          targetPlayer === this.game.myPlayer()
            ? AllPlayers
            : (targetPlayer as PlayerView);
        eventBus.emit(
          new SendEmojiIntentEvent(
            recipient,
            flattenedEmojiTable.indexOf(emoji as Emoji),
          ),
        );
        this.hideTable();
      });
    });
    eventBus.on(CloseViewEvent, (e) => {
      if (!this.hidden) {
        this.hideTable();
      }
    });
  }

  private onEmojiClicked: (emoji: string) => void = () => {};

  private handleBackdropClick = (e: MouseEvent) => {
    const panelContent = this.querySelector(
      'div[class*="bg-zinc-900"]',
    ) as HTMLElement;
    if (panelContent && !panelContent.contains(e.target as Node)) {
      this.hideTable();
    }
  };

  render() {
    if (!this.isVisible) {
      return null;
    }

    return html`
      <div
        class="fixed inset-0 bg-black/15 backdrop-brightness-110 flex items-center justify-center z-[10002]"
        @click=${this.handleBackdropClick}
      >
        <div
          class="bg-zinc-900/95 p-[6px] flex items-center justify-center rounded-[10px] z-[10003] relative shadow-2xl shadow-black/50 ring-1 ring-white/5"
          style="
            width: min(410px, calc(100vw - 60px), calc((100vh - 40px) * 215 / 449));
            aspect-ratio: 215 / 449;
            container-type: size;
          "
          @contextmenu=${(e: MouseEvent) => e.preventDefault()}
          @wheel=${(e: WheelEvent) => e.stopPropagation()}
          @click=${(e: MouseEvent) => e.stopPropagation()}
        >
          <!-- Close button -->
          <button
            class="absolute -top-3 -right-3 w-7 h-7 flex items-center justify-center
                    bg-zinc-700 hover:bg-red-500 text-white rounded-full shadow transition-colors z-[10004]"
            @click=${this.hideTable}
          >
            ✕
          </button>
          <div
            class="flex flex-col"
            style="transform: scale(calc(100cqw / 215px)); transform-origin: center;"
          >
            ${emojiTable.map(
              (row) => html`
                <div class="flex justify-center">
                  ${row.map(
                    (emoji) => html`
                      <button
                        class="flex transition-transform duration-300 ease justify-center items-center cursor-pointer
                                border border-solid border-zinc-600 rounded-[8px] bg-zinc-800 hover:bg-zinc-700 active:bg-zinc-600
                                m-[2px] text-[25px] w-[35px] h-[35px] hover:scale-[1.1] active:scale-[0.95]"
                        @click=${() => this.onEmojiClicked(emoji)}
                      >
                        ${emoji}
                      </button>
                    `,
                  )}
                </div>
              `,
            )}
          </div>
        </div>
      </div>
    `;
  }

  hideTable() {
    this.isVisible = false;
    this.requestUpdate();
  }

  showTable(oneEmojiClicked: (emoji: string) => void) {
    this.onEmojiClicked = oneEmojiClicked;
    this.isVisible = true;
    this.requestUpdate();
  }

  createRenderRoot() {
    return this; // Disable shadow DOM to allow Tailwind styles
  }
}

File: rust/dioxus-ui/src/components/emoji_table.rs
//! EmojiTable component
//!
//! An emoji picker modal that displays a grid of emojis.
//! When an emoji is clicked, it dispatches a custom event with the emoji index.
//! Backdrop clicks and close button dispatch a close event.

use dioxus::prelude::*;
use serde_json::json;
use std::cell::RefCell;

use crate::runtime::emit_ui_event;
use crate::runtime_protocol::{event_keys, event_name};

/// Register the emoji table web component
pub fn register() {
    log::debug!("Registered <dioxus-emoji-table> component");
}

// ── Internal state ───────────────────────────────────────────────

#[derive(Clone, Default)]
struct EmojiTableState {
    is_visible: bool,
    emojis: Vec<String>,
}

// ── Thread-local signals ─────────────────────────────────────────

thread_local! {
    static IS_VISIBLE_SIGNAL: RefCell<Option<Signal<bool>>> = const { RefCell::new(None) };
    static EMOJIS_SIGNAL: RefCell<Option<Signal<Vec<String>>>> = const { RefCell::new(None) };
    static INIT_STATE: RefCell<Option<EmojiTableInitState>> = const { RefCell::new(None) };
}

#[derive(Clone)]
struct EmojiTableInitState {
    emojis: Vec<String>,
}

// ── Event helpers ────────────────────────────────────────────────

fn emit_select(index: usize) {
    emit_ui_event(
        event_name(event_keys::UI_INGAME_EMOJI_TABLE_SELECT),
        Some("component.emoji-table"),
        json!({ "index": index }),
    );
}

fn emit_close() {
    emit_ui_event(
        event_name(event_keys::UI_INGAME_EMOJI_TABLE_CLOSE),
        Some("component.emoji-table"),
        json!({}),
    );
}

// ── Component ────────────────────────────────────────────────────

#[component]
pub fn EmojiTable() -> Element {
    let is_visible = use_signal(|| false);
    let mut emojis = use_signal(|| Vec::<String>::new());

    // Store signals in thread-locals for WASM API access
    IS_VISIBLE_SIGNAL.with(|s| *s.borrow_mut() = Some(is_visible));
    EMOJIS_SIGNAL.with(|s| *s.borrow_mut() = Some(emojis));

    // Initialize emojis from init state on first render
    let init_emojis = INIT_STATE.with(|s| s.borrow().as_ref().map(|state| state.emojis.clone()));
    if let Some(init) = init_emojis {
        if emojis().is_empty() && !init.is_empty() {
            emojis.set(init);
        }
    }

    if !is_visible() {
        return rsx! { div { class: "hidden" } };
    }

    let current_emojis = emojis();

    rsx! {
        // Full-screen backdrop
        div {
            class: "fixed inset-0 bg-black/15 backdrop-brightness-110 flex items-start sm:items-center justify-center z-[10002] pt-4 sm:pt-0",
            onclick: move |_| {
                emit_close();
            },

            div {
                class: "relative",

                // Close button
                button {
                    class: "absolute -top-3 -right-3 w-7 h-7 flex items-center justify-center bg-zinc-700 hover:bg-red-500 text-white rounded-full shadow-sm transition-colors z-[10004]",
                    onclick: move |e| {
                        e.stop_propagation();
                        emit_close();
                    },
                    "\u{2715}"
                }

                // Emoji grid container
                div {
                    class: "bg-zinc-900/95 p-2 sm:p-3 rounded-[10px] z-[10003] shadow-2xl shadow-black/50 ring-1 ring-white/5 w-[calc(100vw-32px)] sm:w-100 max-h-[calc(100vh-60px)] overflow-y-auto",
                    oncontextmenu: move |e| {
                        e.prevent_default();
                    },
                    onclick: move |e| {
                        e.stop_propagation();
                    },

                    // Grid
                    div {
                        class: "grid grid-cols-5 gap-1 sm:gap-2",

                        for (i, emoji) in current_emojis.iter().enumerate() {
                            button {
                                key: "{i}",
                                class: "flex items-center justify-center cursor-pointer aspect-square border border-solid border-zinc-600 rounded-lg bg-zinc-800 hover:bg-zinc-700 active:bg-zinc-600 text-3xl sm:text-4xl transition-transform duration-300 hover:scale-110 active:scale-95",
                                onclick: {
                                    let idx = i;
                                    move |e| {
                                        e.stop_propagation();
                                        emit_select(idx);
                                    }
                                },
                                "{emoji}"
                            }
                        }
                    }
                }
            }
        }
    }
}

// ── Root wrapper ─────────────────────────────────────────────────

fn EmojiTableRoot() -> Element {
    rsx! {
        EmojiTable {}
    }
}

// ── WASM exports ─────────────────────────────────────────────────

pub fn launch_emoji_table(emojis_json: &str) {
    log::info!("Launching emoji table");
    let emojis: Vec<String> = match serde_json::from_str(emojis_json) {
        Ok(e) => e,
        Err(e) => {
            log::error!("Failed to parse emoji list: {}", e);
            Vec::new()
        }
    };
    INIT_STATE.with(|s| {
        *s.borrow_mut() = Some(EmojiTableInitState { emojis });
    });
    let config = dioxus::web::Config::new().rootname("dioxus-emoji-table-root");
    dioxus::LaunchBuilder::new()
        .with_cfg(config)
        .launch(EmojiTableRoot);
}

pub fn show_emoji_table() {
    IS_VISIBLE_SIGNAL.with(|s| {
        if let Some(ref mut sig) = *s.borrow_mut() {
            sig.set(true);
        }
    });
}

pub fn hide_emoji_table() {
    IS_VISIBLE_SIGNAL.with(|s| {
        if let Some(ref mut sig) = *s.borrow_mut() {
            sig.set(false);
        }
    });
}
